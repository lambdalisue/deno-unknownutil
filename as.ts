import type { Predicate } from "./type.ts";
import {
  getPredicateFactoryMetadata,
  setPredicateFactoryMetadata,
  type WithMetadata,
} from "./metadata.ts";
import { isOptional } from "./is.ts";

/**
 * Return an `Optional` annotated type predicate function that returns `true` if the type of `x` is `T` or `undefined`.
 *
 * It is mainly used to annotate fields of `is.ObjectOf()` or parameters of `is.ParametersOf()` that are optional.
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * With `is.ObjectOf()`
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ObjectOf({ foo: as.Optional(is.String) });
 * const a: unknown = { foo: "a" };
 * if (isMyType(a)) {
 *   // a is narrowed to { foo:? string }
 *   const _: { foo?: string } = a;
 * }
 * ```
 *
 * With `is.ParametersOf()`
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ParametersOf([as.Optional(is.String)] as const);
 * const a: unknown = ["a"];
 * if (isMyType(a)) {
 *   // a is narrowed to [string?]
 *   const _: [string?] = a;
 * }
 * ```
 */
function asOptional<T>(
  pred: Predicate<T>,
):
  & Predicate<T | undefined>
  & WithMetadata<AsOptionalMetadata> {
  if (isOptional(pred)) {
    return pred as
      & Predicate<T | undefined>
      & WithMetadata<AsOptionalMetadata>;
  }
  return Object.defineProperties(
    setPredicateFactoryMetadata(
      (x: unknown): x is T | undefined => x === undefined || pred(x),
      { name: "asOptional", args: [pred] },
    ),
    { optional: { value: true as const } },
  ) as
    & Predicate<T | undefined>
    & WithMetadata<AsOptionalMetadata>;
}

export type AsOptionalMetadata = {
  name: "asOptional";
  args: Parameters<typeof asOptional>;
};

/**
 * Return an `Optional` un-annotated type predicate function that returns `true` if the type of `x` is `T`.
 *
 * It is mainly used to annotate fields of `is.ObjectOf()` or parameters of `is.ParametersOf()` that are optional.
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * With `is.ObjectOf()`
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ObjectOf({ foo: as.Required(as.Optional(is.String)) });
 * const a: unknown = { foo: "a" };
 * if (isMyType(a)) {
 *   // a is narrowed to { foo: string }
 *   const _: { foo: string } = a;
 * }
 * ```
 *
 * With `is.ParametersOf()`
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ParametersOf([as.Required(as.Optional(is.String))] as const);
 * const a: unknown = ["a"];
 * if (isMyType(a)) {
 *   // a is narrowed to [string]
 *   const _: [string] = a;
 * }
 */
function asRequired<P extends Predicate<unknown>>(
  pred: P,
): AsRequired<P> {
  if (!isOptional(pred)) return pred as AsRequired<P>;
  const { args } = getPredicateFactoryMetadata(pred);
  return args[0] as AsRequired<P>;
}

export type AsRequired<T> = T extends
  Predicate<undefined | infer U> & WithMetadata<AsOptionalMetadata>
  ? Predicate<U>
  : T extends Predicate<unknown> ? T
  : never;

export const as = {
  Optional: asOptional,
  Required: asRequired,
};
