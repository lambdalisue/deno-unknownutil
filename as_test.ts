import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import type { Equal } from "./_testutil.ts";
import { is } from "./is.ts";
import { as } from "./as.ts";

Deno.test("as.Optional<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, as.Optional(is.Number).name);
    // Nesting does nothing
    await assertSnapshot(
      t,
      as.Optional(as.Optional(is.Number)).name,
    );
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = undefined;
    if (is.ObjectOf({ foo: as.Optional(is.Number) })(a)) {
      assertType<Equal<typeof a, { foo?: number }>>(true);
    }
    if (is.TupleOf([as.Optional(is.Number)] as const)(a)) {
      assertType<Equal<typeof a, [number | undefined]>>(true);
    }
    if (is.ParametersOf([as.Optional(is.Number)] as const)(a)) {
      assertType<Equal<typeof a, [number?]>>(true);
    }
  });
  await t.step("with is.ObjectOf", () => {
    const pred = is.ObjectOf({ foo: as.Optional(is.String) });
    assertEquals(pred({}), true, "omit");
    assertEquals(pred({ foo: undefined }), true, "undefined");
    assertEquals(pred({ foo: "string" }), true, "valid");
    assertEquals(pred({ foo: 0 }), false, "invalid");
  });
  await t.step("with is.TupleOf", () => {
    const pred = is.TupleOf([as.Optional(is.String)]);
    assertEquals(pred([]), false, "omit");
    assertEquals(pred([undefined]), true, "undefined");
    assertEquals(pred(["string"]), true, "valid");
    assertEquals(pred([0]), false, "invalid");
  });
  await t.step("with is.ParametersOf", () => {
    const pred = is.ParametersOf([as.Optional(is.String)]);
    assertEquals(pred([]), true, "omit");
    assertEquals(pred([undefined]), true, "undefined");
    assertEquals(pred(["string"]), true, "valid");
    assertEquals(pred([0]), false, "invalid");
  });
});

Deno.test("as.Required<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(
      t,
      as.Required(as.Optional(is.Number)).name,
    );
    // Non optional does nothing
    await assertSnapshot(t, as.Required(is.Number).name);
    // Nesting does nothing
    await assertSnapshot(
      t,
      as.Required(as.Required(as.Optional(is.Number))).name,
    );
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = undefined;
    if (is.ObjectOf({ foo: as.Required(as.Optional(is.Number)) })(a)) {
      assertType<Equal<typeof a, { foo: number }>>(true);
    }
    if (is.TupleOf([as.Required(as.Optional(is.Number))] as const)(a)) {
      assertType<Equal<typeof a, [number]>>(true);
    }
    if (is.ParametersOf([as.Required(as.Optional(is.Number))] as const)(a)) {
      assertType<Equal<typeof a, [number]>>(true);
    }
  });
  await t.step("with is.ObjectOf", () => {
    const pred = is.ObjectOf({ foo: as.Required(as.Optional(is.String)) });
    assertEquals(pred({}), false, "omit");
    assertEquals(pred({ foo: undefined }), false, "undefined");
    assertEquals(pred({ foo: "string" }), true, "valid");
    assertEquals(pred({ foo: 0 }), false, "invalid");
  });
  await t.step("with is.TupleOf", () => {
    const pred = is.TupleOf([as.Required(as.Optional(is.String))]);
    assertEquals(pred([]), false, "omit");
    assertEquals(pred([undefined]), false, "undefined");
    assertEquals(pred(["string"]), true, "valid");
    assertEquals(pred([0]), false, "invalid");
  });
  await t.step("with is.ParametersOf", () => {
    const pred = is.ParametersOf([as.Required(as.Optional(is.String))]);
    assertEquals(pred([]), false, "omit");
    assertEquals(pred([undefined]), false, "undefined");
    assertEquals(pred(["string"]), true, "valid");
    assertEquals(pred([0]), false, "invalid");
  });
});
