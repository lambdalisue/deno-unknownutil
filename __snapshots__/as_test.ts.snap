export const snapshot = {};

snapshot[`as.Optional<T> > returns properly named function 1`] = `"asOptional(isNumber)"`;

snapshot[`as.Optional<T> > returns properly named function 2`] = `"asOptional(isNumber)"`;

snapshot[`as.Required<T> > returns properly named function 1`] = `"isNumber"`;

snapshot[`as.Required<T> > returns properly named function 2`] = `"isNumber"`;

snapshot[`as.Required<T> > returns properly named function 3`] = `"isNumber"`;
